Improve handling of prePopulated DV

brybacki opened the pull request:
Signed-off-by: Bartosz Rybacki <brybacki@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Previously, when a DataVolume was Reconciled then CDI created PVC
regardless of prePopulated annotation.

There is a possibility that DataVolume and PVC pair is being
restored or moved from other cluster. When DataVolume is marked
as prePopulated, the CDI should not try to populate it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Handle prepopulated DV
```

brybacki commented:
Both tests fail on:
```
"reason": "Error",
"message": "Unable to connect to imageio data source: bad status: 403 Forbidden kubevirt.io/containerized-data-importer/pkg/importer.createImageioReader \tpkg/importer/imageio-datasource.go:589 kubevirt.io/containerized-data-importer/pkg/importer.NewImageioDataSource \tpkg/importer/imageio-datasource.go:71 main.newDataSource \tcmd/cdi-importer/importer.go:235 main.handleImport \tcmd/cdi-importer/importer.go:146 main.main \tcmd/cdi-importer/importer.go:116 runtime.main \tGOROOT/src/runtime/proc.go:255 runtime.goexit \tGOROOT/src/runtime/asm_amd64.s:1581"
}
```

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.22-hpp

mhenriks commented:
Can you give an example of the bug/issue addressed here?  Is it simply that a "pre populated" DV may exist but the PVC may not yet?  What will be the DV status in that case?

mhenriks commented on tests/datavolume_test.go:
```diff
@@ -1119,6 +1119,24 @@ var _ = Describe("[vendor:cnv-qe@redhat.com][level:component]DataVolume tests",
+				dv, _ := f.CdiClient.CdiV1beta1().DataVolumes(dataVolume.Namespace).Get(context.TODO(), dataVolume.Name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				return dv.Status.Phase
+			}, timeout, pollingInterval).Should(Equal(cdiv1.Succeeded))
```
If the PVC does not exist, this should not be Succeeded

brybacki replied:
Thanks for review. 

Yes, I have started this PR before the discussions on the behavior, and realizing that it is better to use real statuses like `Pending ` and `Succeeded`. That is why I made it Work In Progress.

Now I can go back to work and make it production ready.

brybacki replied:
Done, now it is in Pending, and transitions to Succeeded when PVC is available

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.22-upg

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.21-hpp

awels commented:
/test pull-containerized-data-importer-e2e-k8s-1.21-hpp
/test pull-containerized-data-importer-e2e-k8s-1.22-upg
/test pull-containerized-data-importer-e2e-k8s-1.21-ceph

awels commented:
/test pull-containerized-data-importer-e2e-k8s-1.22-hpp

brybacki commented:
lets try If CI is up
/test pull-containerized-data-importer-e2e-k8s-1.22-hpp

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.21-hpp

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.22-upg

mhenriks commented on pkg/controller/datavolume-controller.go:
```diff
@@ -438,88 +438,92 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
-			if err := r.setCloneOfOnPvc(pvc); err != nil {
+		switch selectedCloneStrategy {
+		case HostAssistedClone:
+			if err := r.ensureExtendedToken(pvc); err != nil {
```
I think this PR does a good job of handling the case where a DV has "pre populated" annotation and the PVC does not exist.  But it appears we have to better handle the case where the PVC has "populated for".

This bug has more details:  https://bugzilla.redhat.com/show_bug.cgi?id=2070366

Basically, the controller sees "populated for", takes ownership of the PVC, but still gets to this ` ensureExtendedToken` call

brybacki replied:
do you want me to handle the problem with this PR, or create a new PR?

brybacki replied:
updated the logic, now I need to prepare a test so this can be confirmed,

brybacki replied:
looking at the tests

akalenyu commented:
/retest

brybacki commented:
Updated the logic to cover DV and PVC, also added a test that failed on previous logic on host assisted clone and also on csi clone. 

There is only one problem left. If I try to write a test without adding a `source` PVC it fails on :
```
                Message: "admission webhook \"datavolume-validate.cdi.kubevirt.io\" denied the request:  Source PVC cdi-e2e-tests-dv-func-test-rq7w9/sourcepvcempty not found",
```

brybacki commented:
Ok, that is bad. I know what to do. 
I'll go back to small change, to only handle annotation on DV in this PR, and I'll open a new PR with the changes to handle pvc correctly.

brybacki commented:
@mhenriks I have reverted change here to only handle DV with prePopulated annotation. I would like this to be reviewed and merged. The rest of the problem will be done in a separate PR. 

WIP PR here https://github.com/kubevirt/containerized-data-importer/pull/2227. I am still trying to analyze the case and discover edge cases so it might be a mess.

brybacki commented on pkg/controller/datavolume-conditions.go:
```diff
@@ -122,7 +122,7 @@ func updateReadyCondition(conditions []cdiv1.DataVolumeCondition, status corev1.
 
 func updateBoundCondition(conditions []cdiv1.DataVolumeCondition, pvc *corev1.PersistentVolumeClaim, reason string) []cdiv1.DataVolumeCondition {
-	if pvc != nil {
+	if pvc != nil && pvc.Name != "" {
```
At the beginning of main reconcile loop we do `pvc := &corev1.PersistentVolumeClaim{}` to get the PVC. 
If it is not found we have an empty PVC with no name. 

I can make it null, but I would need to carefully check if all the code is ready for null. This needs to be improved, and I'll try to clean up on the next PR.

mhenriks replied:
Yeah this is pretty ugly, please handle in #2227

mhenriks commented:
/retest

brybacki commented:
/test pull-containerized-data-importer-e2e-k8s-1.22-hpp-istio

brybacki commented:
/cherrypick release-v1.43

brybacki commented:
/cherrypick release-v1.43

aglitke commented:
/cherrypick release-v1.49

aglitke commented:
Oops.  This is already on release-v1.49
